### PR TITLE
feat: no_std

### PR DESCRIPTION
**Pull request by DaniPopes**
## Motivation

Make `ruint` `no_std`-compatible.

Supersedes #260.

## Solution

See original description in #260.
Since `core::error::Error` is unstable, we instead implement the formatting using `core::fmt` and gate `impl std::error::Error` with `#[cfg(feature = "std")]`

## PR Checklist

- [x] Added Tests
- [x] Added Documentation
- [x] Updated the changelog

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -38,30 +38,34 @@ ruint-macro = { version = "1.1.0", path = "ruint-macro" }
 name = "bench_uint"
 harness = false
 path = "benches/bench.rs"
+required-features = ["std"]
 
 [dependencies]
 ruint-macro.workspace = true
 
-thiserror = "1.0"
+thiserror = { version = "1.0", optional = true }
```
Comment (prestwich) [inline Cargo.toml:46]: is this still used anywhere?

File: src/from.rs [hunk 2]
```diff
@@ -32,51 +32,90 @@
...
+
+impl<T> fmt::Display for ToUintError<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::ValueTooLarge(bits, _) => write!(f, "Value is too large for Uint<{bits}>"),
+            Self::ValueNegative(bits, _) => {
+                write!(f, "Negative values can not be represented as Uint<{bits}>")
+            }
+            Self::NotANumber(bits) => {
+                write!(f, "'Not a number' (NaN) not be represented as Uint<{bits}>")
```
Comment (prestwich) [inline src/from.rs:65]: nit: cannot

File: Cargo.toml (hunk 1 above)
Comment (DaniPopes) [reply Cargo.toml:46]: postgres and sql

**Comment by DaniPopes** [comment]
Coverage is actually unchanged because it didn't count code generated by the macro

**Review by prestwich — APPROVED** [review]
lgtm

**Pull request merged**